devicetypes/stelpro/stelpro-maestro-thermostat.src/stelpro-maestro-thermostat.groovy

```diff
@@ -0,0 +1,617 @@
+	//"st wattr 0x${device.deviceNetworkId} 0x19 0x201 0x1C 0x30 {05}"
+}
+
+def setCustomThermostatMode(mode) {
+   setThermostatMode(mode)
+}
+/*
+def on() {
+	//fanOn()
+}
+
+def fanOn() {
+	log.debug "fanOn"
+	sendEvent("name":"thermostatFanMode", "value":"fanOn")
+	"st wattr 0x${device.deviceNetworkId} 0x19 0x202 0 0x30 {04}"
+}
+
+def auto() {
+	fanAuto()
+}
+
+def fanAuto() {
+	log.debug "fanAuto"
+	sendEvent("name":"thermostatFanMode", "value":"fanAuto")
+	//"st wattr 0x${device.deviceNetworkId} 1 0x202 0 0x30 {05}"
+}
+*/
+def configure() {
+	log.debug "binding to Thermostat cluster"
+	delayBetween([
+		sendEvent("name":"thermostatMode", "value":"heat"),
+		"zdo bind 0x${device.deviceNetworkId} 1 0x19 0x201 {${device.zigbeeId}} {}",
+		//Cluster ID (0x0201 = Thermostat Cluster), Attribute ID, Data Type, Payload (Min report, Max report, On change trigger)
+		zigbee.configureReporting(0x0201, 0x0000, 0x29, 10, 60, 50), 	//Attribute ID 0x0000 = local temperature, Data Type: S16BIT
+		zigbee.configureReporting(0x0201, 0x0012, 0x29, 1, 0, 50),  	//Attribute ID 0x0012 = occupied heat setpoint, Data Type: S16BIT
+		zigbee.configureReporting(0x0201, 0x0008, 0x20, 300, 900, 5),   //Attribute ID 0x0008 = pi heating demand, Data Type: U8BIT
+		
+		//Cluster ID (0x0204 = Thermostat Ui Conf Cluster), Attribute ID, Data Type, Payload (Min report, Max report, On change trigger)
+		zigbee.configureReporting(0x0204, 0x0000, 0x30, 1, 0, 1),   //Attribute ID 0x0000 = temperature display mode, Data Type: 8 bits enum
+		zigbee.configureReporting(0x0204, 0x0001, 0x30, 1, 0, 1),   //Attribute ID 0x0001 = keypad lockout, Data Type: 8 bits enum
+		
+		zigbee.configureReporting(0x0405, 0x0000, 0x21, 10, 300, 1), 		//Attribute ID 0x0000 = local humidity, Data Type: U16BIT
+		
+		//Read the configured variables
+		zigbee.readAttribute(0x201, 0x0000),	//Read Local Temperature
+		zigbee.readAttribute(0x201, 0x0012),	//Read Heat Setpoint
+		zigbee.readAttribute(0x201, 0x0008),	//Read PI Heating State
+		zigbee.readAttribute(0x204, 0x0000),	//Read Temperature Display Mode
+		zigbee.readAttribute(0x204, 0x0001),	//Read Keypad Lockout
+		zigbee.readAttribute(0x405, 0x0000),	//Read Local Humidity
+	], 200)
+}
+
+def updated() {
+	response(parameterSetting())
+}
+
+def parameterSetting() {
+	def lockmode = null
+	def valid_lock = 0
+
+	log.info "lock : $settings.lock"
+	if (settings.lock == "Yes") {
+		lockmode = 0x01
+		valid_lock = 1
+	}
+	else if (settings.lock == "No") {
+		lockmode = 0x00
+		valid_lock = 1
+	}
+	
+	if (valid_lock == 1)
+	{
+		log.info "lock valid"
+		delayBetween([
+			zigbee.writeAttribute(0x204, 0x01, 0x30, lockmode),	//Write Lock Mode
+			poll(),
+		], 200)
+	}
+	else {
+		log.info "nothing valid"
+	}
+}
+
+private hex(value) {
+	new BigInteger(Math.round(value).toString()).toString(16)
+}
+
+private getEndpointId() {
+	new BigInteger(device.endpointId, 16).toString()
+}
+
+private String swapEndianHex(String hex) {
```

Comment:
There is a `zigbee.swapEndianHex` that you can use instead

Reply:
Good, at the time, this was a custom method found on the forum.
Donald Kirker wrote me this 2 years ago:
> "One gotcha is the value in the global write command needs to be little endian formatted, so if the desired value is 0xABCD you need to pass `{CDAB}`."

> So some of the values may be little endian.

Reply:
Yeah, we unfortunately didn't have a standard library then. Thankfully around that time efforts evolved to get all of these re-used functions together!